src/base64.cpp

```diff
@@ -14,36 +14,152 @@
+    unsigned long value = 0;
+    uint8_t code[4] = {'=', '=', '=', '='};
+    switch (size) {
+        case 1:
+            value = input[0] << 4;
+            code[1] = b64base[value & 0x3f];
+            value >>= 6;
+            code[0] = b64base[value];
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+            sout << std::string_view(reinterpret_cast<char *>(code), 4);
+            break;
+        case 2:
+            value = input[0];
+            value <<= 8;
+            value |= input[1];
+            value <<= 2;
+            code[2] = b64base[value & 0x3f];
+            value >>= 6;
+            code[1] = b64base[value & 0x3f];
+            value >>= 6;
+            code[0] = b64base[value];
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+            sout << std::string_view(reinterpret_cast<char *>(code), 4);
+            break;
+        default:
+            break;
+    }
+    return 0;
+}
+
+// Acumulates bytes in input buffer until 4 bytes are available.
+// Translate the 4 bytes from Base64 form and append to buffer.
+// Returns new number of bytes in buffer.
+static size_t b64decode(uint8_t c, uint8_t *input, size_t size, std::ostringstream &sout) {
+    // ignore invalid characters
+    if (b64unbase[c] > 64) {
+        return size;
+    }
+    input[size++] = c;
+    // decode atom
+    if (size == 4) {
+        uint8_t decoded[3];
+        int valid = 0;
+        int value = 0;
+        value = b64unbase[input[0]];
+        value <<= 6;
+        value |= b64unbase[input[1]];
+        value <<= 6;
+        value |= b64unbase[input[2]];
+        value <<= 6;
+        value |= b64unbase[input[3]];
+        decoded[2] = static_cast<uint8_t>(value & 0xff);
+        value >>= 8;
+        decoded[1] = static_cast<uint8_t>(value & 0xff);
+        value >>= 8;
+        decoded[0] = static_cast<uint8_t>(value);
+        // take care of paddding
+        valid = (input[2] == '=') ? 1 : (input[3] == '=') ? 2 : 3;
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        sout << std::string_view(reinterpret_cast<char *>(decoded), valid);
+        return 0;
+        // need more data
+    } else {
+        return size;
+    }
+}
 
 std::string encode_base64(const std::string &input) {
-    std::istringstream sin(input);
     std::ostringstream sout;
-    base64::encoder E; // NOLINT(clang-analyzer-optin.cplusplus.UninitializedObject)
-    E.encode(sin, sout);
+    uint8_t ctx[4]{};
+    size_t ctxlen = 0;
+    size_t ncols = 0;
+    for (const char b : input) {
+        ctxlen = b64encode(static_cast<uint8_t>(b), ctx, ctxlen, sout);
+        if (ctxlen == 0) { // appended 4 characters
+            ncols += 4;
+            if (ncols >= 72) { // add CRLF every 72 columns
+                sout << "\r\n";
+                ncols = 0;
+            }
+        }
+    }
+    b64pad(ctx, ctxlen, sout);
```

Comment:
Is there a way for the padding to blow up the 72 col limit? I think so, right? So perhaps add another if before padding?

Reply:
I don't think this is possible (I tried), I don't want to add a check for this if it's a branch never taken.

Reply:
It was a good question, but I agree it is not possible. The padding always rounds up the number of bytes in the output to a multiple of 4. Since 72 is also a multiple of 4, the padding can never be the reason a line goes over the limit.